components/layout_2020/fragments.rs

```diff
@@ -342,3 +346,37 @@ impl CollapsedMargin {
         self.max_positive + self.min_negative
     }
 }
+
+impl Serialize for BoxFragment {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut serializer = serializer.serialize_struct("BoxFragment", 6)?;
+        serializer.serialize_field("content_rect", &self.content_rect)?;
+        serializer.serialize_field("padding", &self.padding)?;
+        serializer.serialize_field("border", &self.border)?;
+        serializer.serialize_field("margin", &self.margin)?;
+        serializer.serialize_field(
+            "block_margins_collapsed_with_children",
+            &self.block_margins_collapsed_with_children,
+        )?;
+        serializer.serialize_field("children", &self.children)?;
+        serializer.end()
+    }
+}
+
+impl Serialize for TextFragment {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut serializer = serializer.serialize_struct("TextFragment", 3)?;
+        serializer.serialize_field("rect", &self.rect)?;
+        serializer.serialize_field("ascent", &self.ascent)?;
+        serializer.serialize_field("glyphs", &self.glyphs)?;
+        serializer.end()
+    }
+}
+
+impl Serialize for ImageFragment {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut serializer = serializer.serialize_struct("ImageFragment", 1)?;
+        serializer.serialize_field("rect", &self.rect)?;
+        serializer.end()
+    }
+}
```

Comment:
For these three impls: why are they not derived? Could they be derived with some attributes from https://serde.rs/field-attrs.html ?

Reply:
I think I learnt about `#[serde(skip_serializing)]` after implementing the serializers. Fixed.